### ISSUE TITLE
Use Libhybris to support Wayland and more native GPU acceleration

### ISSUE THREAD

**Issue by samzanemesis** [maruos/maruos#38]
Libhybris is a compatibility layer that overrides bionic symbols for GNU based systems, in layman's terms it can you to use Android drivers and features in common linux distros

https://github.com/libhybris/libhybris

This project would greately benefit if had Hybris implemented into the system by allowing to use accelerated Wayland & etc

**Comment by bootlessxfly** [comment]
Is libhybris not already used in this framework. How do we access 3d support in the Linux userland of it is not used?

**Comment by pdsouza** [comment]
Thanks for the suggestion! I would REALLY like to have hardware acceleration in Maru OS.

**Comment by poVoq** [comment]
This thread discusses a similar setup as MaruOS on the Zenfone2:
http://forum.xda-developers.com/zenfone2/general/zf2-linux-howto-t3158034
While not perfect, they got a pretty well running OpenGL software implementation running on the CPU via LLVM. Use of Libhybis is also discussed.

**Comment by curiousercreative** [comment]
+1 for hardware acceleration!

**Comment by curiousercreative** [comment]
@pdsouza have you or another dev begun any development toward hardware acceleration? I became curious after getting the Franco kernel working and I've been able to get libhybris built and installed on the Debian OS, but I believe the host OS needs a pre-build patch and I haven't got a Maru build environment just yet. In case this is at all helpful, I documented how I got libhybris installed on our Debian [here](https://github.com/libhybris/libhybris/wiki/Building). In addition, I believe [a patch similar to this](http://mrs.ro/dl/odroid/bionic.patch) is needed for our host OS before our guest Debian can access the hardware. I don't much know what else would be required, but I'm happy to pitch in and learn.

**Comment by pdsouza** [comment]
@curiousercreative Oh wow, great work! No, no one I know of in the community who's working on hardware acceleration right now. I myself am not familiar with libhybris at all so haven't looked at this at all. Your documentation on how to build it is very helpful though! I am VERY interested in getting hardware acceleration to work on Maru, so am happy to answer any questions you have as you dig through. If I can get some extra bandwidth before you get your Maru build environment all set up, I can try applying the patch and looking at this further, but I've been pretty booked recently. I still am very hazy on how libhybris works (how does Debian hook into the Android graphics libraries? Do we need to mount the Android graphics libraries into the Debian container somewhere? Do we need to install additional software on Debian for the display server to become libhybris-aware? I guess I should just get in touch with the libhybris folks on GitHub haha...) so if you find any useful resources on how it works as you tinker around, please do let me know! Please feel free to open up a thread on our dev forum to discuss this in case anyone else is interested in helping out too.

**Comment by curiousercreative** [comment]
I could be way off here as I'm not a Linux or Android developer, but it seems that hardware acceleration can be broken up into at least 3 pieces:
1. Debian graphics API adapter for Android hardware
  [Libhybris](https://github.com/libhybris/libhybris) seems to be what is necessary. We've had some success installing this on Debian.
1. Host Android hardware made accessible to guest Debian
  Because Debian is running as a guest OS within a container, we'll need to give it access to the hardware drivers. We could probably get this working, but needs to be built into Maru.
1. Debian desktop environment and window manager that support hardware acceleration
  This is where I'm haziest. I'm unsure as to whether our xfce can be feasibly hardware accelerated or whether we would need to consider using a different desktop environment altogether.

An outstanding question: do we need to be using Wayland for any of this to work? I ask because of this quote from the [libhybris wikipedia article](https://en.wikipedia.org/wiki/Hybris_(software)):
>Hybris can also translate Android’s EGL calls into Wayland EGL calls, allowing Android graphic drivers to be used on Wayland-based systems

Dropping some links here for myself and others:
- https://wiki.debian.org/DesktopEnvironment
- https://wiki.gentoo.org/wiki/Xorg/Hardware_3D_acceleration_guide
- https://wayland.freedesktop.org/architecture.html
- https://plasma-mobile.org/overview/#panel-269-2-0-0
- https://en.wikipedia.org/wiki/Adreno
- https://en.wikipedia.org/wiki/Hybris_(software)

**Comment by bootlessxfly** [comment]
Currently maru is running on xserver. I think you we will also need to get wayland running in maru. KDE Plasma can run in wayland, it is a little buggy though. I believe KDE plasma wayland can be installed in Debian Sid. I have not had a chance to try this out yet, however.

**Comment by luka177** [comment]
We can keep Xorg for now we can use xorg with libhybris

**Comment by luka177** [comment]
It sounds like topic is discounted, but still current status:
Compositors that have hwcomposer backend: kwin (plasma and plasma mobile hwcomposer1 and hwc2 wip), wlroots(sway and phosh hwc1 and hwc2), mir(ubuntu touch hwc1 and hwc2), xorg(hwc1 and hwc2). Most work is done for android 7 and 9, must work on 8 too. No android 10 support yet.

**Comment by twaik** [comment]
Looks like this issue is stale and can be closed...
Using libhybris for hardware acceleration will require patching bionic with halium patches (i.e. because Android's GLES uses thread-local-storage slots which are reserved for other programs in glibc). But you can try use android-emugl for this (I can provide some sources for desktop linux). It is impossible to draw directly to Xserver with GLES, but it is possible to draw to pbuffer surface and then read pixels directly to XImage...
